packages/fxa-auth-server/lib/db/index.ts

```diff
@@ -0,0 +1,1126 @@
+
+      if (
+        !this.redis ||
+        !TOKEN_PRUNING_ENABLED ||
+        !features.isLastAccessTimeEnabledForUser(uid)
+      ) {
+        return;
+      }
+
+      const tokenIds = sessionTokens
+        .filter(
+          (token: any) => token.createdAt <= Date.now() - TOKEN_PRUNING_MAX_AGE
+        )
+        .map((token: any) => token.id);
+
+      if (tokenIds.length === 0) {
+        return;
+      }
+
+      return this.redis.pruneSessionTokens(uid, tokenIds);
+    }
+
+    async createDevice(uid: string, deviceInfo: any): Promise<any> {
+      log.trace('DB.createDevice', { uid: uid, id: deviceInfo.id });
+      const sessionTokenId = deviceInfo.sessionTokenId;
+      const refreshTokenId = deviceInfo.refreshTokenId;
+
+      const id = await random.hex(16);
+      deviceInfo.id = id;
+      deviceInfo.createdAt = Date.now();
+
+      try {
+        await Device.create({
+          ...deviceInfo,
+          sessionTokenId,
+          refreshTokenId,
+          uid,
+        });
+      } catch (err) {
+        if (isRecordAlreadyExistsError(err)) {
+          const devices = await this.devices(uid);
+          // It's possible (but extraordinarily improbable) that we generated
+          // a duplicate device id, so check the devices for this account. If
+          // we find a duplicate, retry with a new id. If we don't find one,
+          // the problem was caused by the unique sessionToken or
+          // refreshToken constraint so return an appropriate error.
+          const duplicateDevice = devices.find(
+            (device: any) => device.id === deviceInfo.id
+          );
+          if (duplicateDevice) {
+            return this.createDevice(uid, deviceInfo);
+          }
+
+          const conflictingDevice = devices.find(
+            (device: any) =>
+              (sessionTokenId && device.sessionTokenId === sessionTokenId) ||
+              (refreshTokenId && device.refreshTokenId === refreshTokenId)
+          );
+          throw error.deviceSessionConflict(conflictingDevice?.id);
+        }
+        throw err;
+      }
+      deviceInfo.pushEndpointExpired = false;
+      return deviceInfo;
+    }
+
+    async updateDevice(uid: string, deviceInfo: any) {
+      const sessionTokenId = deviceInfo.sessionTokenId;
+      try {
+        await Device.update({
+          uid,
+          callbackIsExpired: deviceInfo.pushEndpointExpired,
+          ...deviceInfo,
+        });
+      } catch (err) {
+        if (isNotFoundError(err)) {
+          throw error.unknownDevice();
+        }
+        if (isRecordAlreadyExistsError(err)) {
+          // Identify the conflicting device in the error response,
+          // to save a server round-trip for the client.
+          const devices = await this.devices(uid);
+          const conflictingDevice = devices.find(
+            (device: any) => device.sessionTokenId === sessionTokenId
+          );
+          throw error.deviceSessionConflict(conflictingDevice.id);
+        }
```

Comment:
I'm getting a warning that conflictingDevice is possibly undefined

```suggestion
          throw error.deviceSessionConflict(conflictingDevice?.id);
```

Reply:
I just pushed a change for this, cause I was curious if let the CI pass.